src/test/unit_tests/test_s3select.js

```diff
@@ -0,0 +1,201 @@
+const { Transform } = require('readable-stream');
+const stream_utils = require('../../util/stream_utils');
+
+const csv_small_str =
+`a,b,c,d,e,
+0,414,339,155,67,
+1,60,741,755,698,
+2,572,53,375,241,
+3,966,233,168,371,
+4,858,55,964,901,
+5,630,524,728,489,
+6,355,260,504,32,
+7,636,736,662,50,
+8,75,169,470,487,
+9,910,225,185,834,
+`;
+
+const json_str =
+`[
+{"doublef": 3.3, "intf" : 3, "stringf": "i am a string"},
+{"doublef": 3.3, "intf" : 123, "stringf": "i am a string"}
+]
+`;
+
+/*
+Handles streaming an input (csv/json) and getting the sql result as output.
+The stream in src/util/s3select.js also encodes the result into chunks according to AWS-defined format,
+which is not suitable here.
+(If we do want to reuse rc/util/s3select.js, it's possible to factor out the encoding, eg add a parameter or add a stream
+to handle encoding specifically.)
+*/
+class S3SelectStream extends Transform {
+
+    constructor(context) {
+        super({});
+        this.s3select = new (nb_native.S3Select)(context);
+    }
+
+    async _transform(chunk, encoding, cb) {
+        //console.log("got chunk ", chunk.length);
+        let select = await this.s3select.write(chunk);
+        if (select) {
+            this.push(select.select);
+        }
+        return cb();
+    }
+
+    async _flush(cb) {
+        //console.log("in flush");
+        let select = await this.s3select.flush();
+        if (select) {
+            this.push(select.select);
+        }
+        return cb();
+    }
+}
+
+class CollectChunks extends stream.Writable {
+
+    chunks = [];
+
+    _write(chunk, enc, next) {
+        //console.log("got chunk");
+        this.chunks.push(chunk);
+        next();
+    }
+
+}
+
+class FailingTransform extends Transform {
+
+    constructor(writable) {
+        super({});
+        this.writable = writable;
+    }
+
+    _transform(chunk, enconding, cb) {
+        this.writable.write(chunk);
+        return cb();
+    }
+
+    _flush(cb) {
+        if (this.writable.flush) {
+            this.writable.flush();
+        }
+        return cb();
+    }
+
+}
+
+/*class FailingReadble extends Readable {
+
+}*/
+
+async function run_sql(args, input_str) {
+    const s3select = new S3SelectStream(args);
+    const input_stream = stream.Readable.from([input_str]);
+    const collect = new CollectChunks();
+    await stream.promises.pipeline(input_stream, s3select, collect);
+    return Buffer.concat(collect.chunks).toString();
+}
+
+async function simulate_async_loop(writable) {
+    writable.write("this is an ");
+    await new Promise(resolve => {
+        writable.write("input string.", null, () => {
+            resolve();
+        });
+    });
+    assert(writable.emit('error', new Error("I am an error event in an async loop.")));
+}
+
+
+mocha.describe('s3select', function() {
+
+    mocha.it('should create ok', function() {
+        const select_args = {
+            query: "select sum(int(_2)) from stdin;",
```

Comment:
Would you please add a comment line before the query with a human-readable query?
For example, this query sums the numbers in the second column (one-liner, easy to read).